### PR TITLE
[IA-456] org units tree view

### PR DESCRIPTION
**Pull request by geoffreydhuyvetters**
## Changes

- Created `IasoTreeView` based on MaterialUI's `Treeview`
- Created `DynamicSearch`based on Iaso's `OrgUnitSearch`
- Combined both into `TreeviewWithSearch`, which allows to select an item in the tree through a text search
- Integrated `TreeviewWithSearch` into `OrgUnitTreeviewModal`, which handles all parameters specific to Iaso
- Added missing `onCancel` callback in `CreateReAssignDialogComponent`
- Removed `OrgUnitSearch` and replaced it with `TreeviewWithModal` in : 
    - `CreateReAssignDialogComponent`
    - `UserLocations`
    - `InstancesFiltersComponent`
- Removed call to `/orgunits/` in Instances test, as the call is now made by the Treeview
- The tooltip that appeared on hover have been removed, since the treeview show the hierarchy, and the source used is the default source

## Features

- Selecting an OrgUnit via the search box will expand its parents and highlight it
- Pressing `Cancel` will restore previous selection
- Selected orgUnits are displayed in the form of a "truncated" treeview when the modal is closed
- Selection can be cleared in the truncated treeview view
- Choice between single- or multiselect
- When selecting from the search box, the Treeview will scroll to the selected item

## Points of attention for the review
- There's a lot of data manipulation, so there's certainly ways to improve performance, and a code smell here or there
- Naming is pretty terrible overall
- There's certainly some `useCallback`s that are superfluous and should be removed

## Known issues
- in the instances screen, navigating back after a search will not clear the `TruncatedTreeview` 
- in `CreateReAssignDialogComponent`, when an org unit is pre-selected, the tree expands to show it, but does not highlight it

### Single select

https://user-images.githubusercontent.com/38907762/126803807-ab806996-3574-4375-88b6-68e6c64a20e1.mov

## Multi select

https://user-images.githubusercontent.com/38907762/126804212-d9814fb1-c0a6-4e2d-980b-595bc3ddedab.mov

Issue: https://bluesquare.atlassian.net/browse/IA-456

### CODE REVIEW

**Comment by olethanh** [comment]
converting it back to draft, as I have fix to do in the API before merging and don't want it merged by mistake

**Review by olethanh — CHANGES_REQUESTED** [review]

File: hat/assets/js/apps/Iaso/domains/instances/components/InstancesFiltersComponent.js [hunk 1]
```diff
@@ -103,6 +104,15 @@ class InstancesFiltersComponent extends Component {
         redirectToReplace(baseUrl, tempParams);
     }
 
+    // eslint-disable-next-line class-methods-use-this
+    tooltip(orgUnit, icon) {
```
Comment (olethanh) [inline hat/assets/js/apps/Iaso/domains/instances/components/InstancesFiltersComponent.js:108]: we don't seem to be using it?

File: hat/assets/js/apps/Iaso/domains/orgUnits/components/TreeView/requests.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+};
+
+const getRootData = async () => {
+    const response = await iasoGetRequest({
+        disableSuccessSnackBar: true,
+        requestParams: {
+            url: `/api/orgunits/?&rootsForUser=true&defaultVersion=true&validation_status=all${LIMIT_HACK}`,
+        },
+    });
+    const useableData = response.orgunits.map(orgUnit => {
```
[the remark this answers is not in the document]
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/orgUnits/components/TreeView/requests.js:28]: That was the data I actually used. Granted, the name was made moot when I put the whole orgUnit in `data`

File: hat/assets/js/apps/Iaso/domains/instances/components/InstancesFiltersComponent.js (hunk 1 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/instances/components/InstancesFiltersComponent.js]: indeed, it seems to be a leftover

**Comment by quang-le** [comment]
## Updates 10/08/2021

- removed `OrgUnitLevelFilterComponent`and `OrgUnitsLevelsFiltersComponent`
- replaced them with `OrgUnitTreeViewModal` in `OrgUnitFiltersComponent` and `OrgUnitInfosComponent`
- updated `requests.js` and `OrgUnitTreeViewModal` to enable changing the source

**Comment by quang-le** [comment]
## Updates 11/08/2021

- Add styling: resize Input box to match other fields in OrgUnits list, correctly align `TruncatedTreeview`, change box color on hover, align box with siblings
- Change "clear" Icon
- Clear OrgUnit selection when changing source
- Restore deep linking in OrgUnit List
- Update deps to bluesquare-components and node-sass



https://user-images.githubusercontent.com/38907762/129043156-23cee63c-3b36-4bc5-945a-e2b81d96387a.mov

**Comment by beygorghor** [comment]
We need to add `ignoreEmptyNames` tu api calls to avoid weird results
Screenshot 2021-08-12 at 14 43 29
Selected org unit is not visible in the tree.
Also maybe put a max height to the dialog in order to avoid this double scroll

File: hat/assets/js/apps/Iaso/domains/orgUnits/components/TreeView/OrgUnitTreeviewPicker.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import { Paper } from '@material-ui/core';
+import { makeStyles } from '@material-ui/core/styles';
+import { IconButton, useSafeIntl } from 'bluesquare-components';
+import { MESSAGES } from './messages';
+import { TruncatedTreeview } from './TruncatedTreeview';
+
+const styles = theme => ({
+    placeholder: {
+        alignItems: 'center',
+        fontSize: '16px',
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/orgUnits/components/TreeView/OrgUnitTreeviewPicker.js:12]: fyi you can only use fontSize: 16

**Review by beygorghor — CHANGES_REQUESTED** [review]
Ok just change source while editing an org unit like we said

File: hat/assets/js/apps/Iaso/domains/instances/components/InstancesFiltersComponent.js [hunk 4]
```diff
@@ -113,16 +110,20 @@ const InstancesFiltersComponent = ({
...
-        const parentIds = getOrgUnitParentsIds(orgUnit);
-        parentIds.push(orgUnit.id);
-        const tempParams = {
-            ...params,
-            levels: parentIds.join(','),
-        };
-
-        dispatch(redirectTo(baseUrl, tempParams));
-        dispatch(setInstancesFilterUpdated(true));
+    // FIXME: Olivier check after merge
```
[the remark this answers is not in the document]
Comment (olethanh) [reply hat/assets/js/apps/Iaso/domains/instances/components/InstancesFiltersComponent.js:113]: yep

File: hat/assets/js/apps/Iaso/domains/orgUnits/components/TreeView/OrgUnitTreeviewPicker.js (hunk 3 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/orgUnits/components/TreeView/OrgUnitTreeviewPicker.js:12]: Old habits die hard 🙂

File: hat/assets/js/apps/Iaso/domains/orgUnits/components/TreeView/OrgUnitTreeviewModal.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+import { bool, func, object, arrayOf, oneOfType } from 'prop-types';
+import { TreeViewWithSearch } from './TreeViewWithSearch';
+import ConfirmCancelDialogComponent from '../../../../components/dialogs/ConfirmCancelDialogComponent';
+import { MESSAGES } from './messages';
+import { getRootData, getChildrenData, searchOrgUnits } from './requests';
+import { OrgUnitLabel, getOrgUnitAncestors } from '../../utils';
+import OrgUnitTooltip from '../OrgUnitTooltip';
+import { OrgUnitTreeviewPicker } from './OrgUnitTreeviewPicker';
+
+const tooltip = (orgUnit, icon) => (
```
[the remark this answers is not in the document]
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/orgUnits/components/TreeView/OrgUnitTreeviewModal.js]: Done already

**Comment by quang-le** [comment]
## Update 16/08/2021
- Added Treeview in `OrgUnitInfosComponent`
- Updated messages and added translations
- Made placeholder configurable for `TreeviewPicker`


https://user-images.githubusercontent.com/38907762/129550621-145c3622-23c4-47a9-9d1c-62d721a75529.mov

**Review by beygorghor — APPROVED** [review]

**Pull request merged**